src/common/bmpbndl.cpp

```diff
@@ -292,18 +292,35 @@ wxBitmap wxBitmapBundleImplSet::GetBitmap(const wxSize& size)
 
     // We only get here if the requested size is larger than the size of all
     // the bitmaps we have, in which case we have no choice but to upscale one
-    // of the bitmaps, so find the largest available non-generated bitmap.
-    for ( size_t i = n; n > 0; --i )
+    // of the bitmaps, so find the largest available non-generated bitmap which
+    // can be scaled using an integer factor or, failing that, just the largest
+    // non-generated bitmap.
+    const Entry* entryToRescale = NULL;
+    const Entry* entryLastNotGen = NULL;
+    for ( size_t i = 0; i < n; ++i )
     {
-        const Entry& entry = m_entries[i - 1];
-        if ( !entry.generated )
-        {
-            const Entry entryNew(entry, size);
+        const Entry& entry = m_entries[i];
+        if ( entry.generated )
+            continue;
 
-            m_entries.push_back(entryNew);
+        entryLastNotGen = &entry;
 
-            return entryNew.bitmap;
-        }
+        const double
+            scale = static_cast<double>(size.y) / entry.bitmap.GetSize().y;
+        if ( scale == wxRound(scale) )
+            entryToRescale = &entry;
+    }
+
+    if ( !entryToRescale )
+        entryToRescale = entryLastNotGen;
+
+    if ( entryToRescale )
+    {
+        const Entry entryNew(*entryToRescale, size);
+
+        m_entries.push_back(entryNew);
+
+        return entryNew.bitmap;
```

Comment:
The contract of `GetBitmap(size)` is that it will return the bitmap of exactly this size. So if it's called with a size which can't be obtained by integer scaling, it will scale the bitmap using non-integer factor. I think it might be actually better to do what `wxArtProvider::RescaleOrResizeIfNeeded()` does in this case, see https://github.com/wxWidgets/wxWidgets/blob/73d425b34285b1ba150a1748c2b9146d58125e09/src/common/artprov.cpp#L327,L333 but it will still do something to ensure that the bitmap has the requested size.

However normally only integer scaling should be used because only the bitmaps of "preferred" size should be requested. There are some exceptions, however. The one I've already encountered is that `wxToolBar` with a fixed tool bitmap size will always request bitmaps of this size. I'm not sure what to do about this, so far I've documented (in a a not yet pushed commit) that using `SetToolBitmapSize()` is just not recommended and that you should let toolbar adapt to the size of the bitmaps it has.